### DISCUSSION TITLE
Unable to import _C from detectron2 inside Docker container

### DISCUSSION THREAD

**Discussion by alwibowo** [facebookresearch/detectron2#5345] [Q&A]
Hello,

I have this weird import error when trying to install Detectron2 inside a docker container.

### Context

I basically built detectron2 from the source and pushed it into a private repository in packagecloud. I built it by using the following specific python and torch versions:

- python: 3.10.13
- torch: 2.2.1

### Problem

This approach actually works well locally. I currently use M1 Pro MacOs and manage to install detectron2 from the packagecloud. My model runs without any problem. But when I try to install it inside docker container, this is where the problem starts. I encounter the following error:

```python
ImportError: /app/.venv/lib/python3.10/site-packages/detectron2/_C.cpython-310-aarch64-linux-gnu.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb
```
I tried to import _C from detectron2 locally on my M1 Pro and it worked.  Has anyone encountered this before and managed to make it work in a docker container?

**Comment by AbhishekKallolimath** [comment]
Hi,

The important clue here is:

```text
detectron2/_C.cpython-310-aarch64-linux-gnu.so:
undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb
````

This is a C++ symbol from PyTorch, so the problem is most likely not the Python package itself. It indicates that the Detectron2 native extension (`_C.so`) was compiled against a different PyTorch binary than the one that is being loaded inside the Docker container.

That also explains why it can work on the M1 host but fail in Docker: the host and container are separate environments, and the Detectron2 extension must be compiled for the exact Linux/PyTorch environment in which it will run.

### 1. Check the environment inside Docker

Before rebuilding anything, run:

```bash
uname -m

python -c "import platform; print(platform.platform())"
python -c "import sys; print(sys.version)"

python -c "import torch; print('torch:', torch.__version__); print('torch path:', torch.__file__)"

python -c "import torchvision; print('torchvision:', torchvision.__version__); print('torchvision path:', torchvision.__file__)"

python -m pip show detectron2

python -m detectron2.utils.collect_env
```

The first command is particularly important because your failing extension is:

```text
aarch64-linux-gnu.so
```

so the container is using Linux ARM64.

Also make sure the PyTorch and torchvision versions actually match. For the versions you mentioned, PyTorch 2.2.1 is paired with torchvision 0.17.1. PyTorch's version matrix confirms that pairing for Linux. ([[PyTorch Documentation](https://docs.pytorch.org/get-started/previous-versions/?utm_source=chatgpt.com)][2])

### 2. Install PyTorch first inside the container

Do not install Detectron2 first and then change PyTorch afterwards.

For example, if you specifically need PyTorch 2.2.1, install the matching torchvision version first:

```bash
python -m pip install --upgrade pip

python -m pip install --no-cache-dir \
    torch==2.2.1 \
    torchvision==0.17.1
```

If you are using a specific CPU/CUDA build, use the corresponding PyTorch index instead of blindly mixing wheels. PyTorch publishes separate builds for CPU/CUDA/other platforms. ([[PyTorch Documentation](https://docs.pytorch.org/get-started/previous-versions/?utm_source=chatgpt.com)][2])

Then verify:

```bash
python -c "import torch, torchvision; print(torch.__version__); print(torchvision.__version__)"
```

You should see the expected pair:

```text
2.2.1
0.17.1
```

### 3. Remove the existing Detectron2 binary

If Detectron2 was installed from your package repository, uninstall it before rebuilding:

```bash
python -m pip uninstall -y detectron2
```

If you have a local Detectron2 checkout, also remove old compiled artifacts:

```bash
rm -rf build
find . -name "*.so" -delete
```

Detectron2's documentation specifically recommends removing `build/` and previously generated `.so` files before rebuilding when PyTorch has changed. ([[GitHub](https://github.com/facebookresearch/detectron2/blob/main/INSTALL.md?utm_source=chatgpt.com)][1])

### 4. Build Detectron2 inside the Docker container

The safest approach is to compile Detectron2 **inside the same container** after PyTorch has already been installed:

```bash
git clone https://github.com/facebookresearch/detectron2.git
cd detectron2

python -m pip install --no-cache-dir --no-build-isolation -e .
```

The important part is that the build happens after the container's final PyTorch installation.

Alternatively, without cloning:

```bash
python -m pip install --no-cache-dir --no-build-isolation \
    'git+https://github.com/facebookresearch/detectron2.git'
```

Detectron2 officially supports building from source this way. ([[GitHub](https://github.com/facebookresearch/detectron2/blob/main/INSTALL.md?utm_source=chatgpt.com)][1])

### 5. Test the compiled extension immediately

Before running your application, test:

```bash
python -c "import torch; import detectron2; from detectron2 import _C; print('Detectron2 _C imported successfully')"
```

If that succeeds, the specific `_C` loading problem is fixed.

You can also run:

```bash
python -m detectron2.utils.collect_env
```

again and keep the output for comparison.

### 6. Make the Docker build reproducible

I would structure the Dockerfile so that the dependency order is explicit:

```dockerfile
FROM python:3.10

RUN apt-get update && apt-get install -y \
    git \
    gcc \
    g++ \
    && rm -rf /var/lib/apt/lists/*

WORKDIR /app

RUN python -m pip install --upgrade pip

# Install the exact PyTorch stack first
RUN python -m pip install --no-cache-dir \
    torch==2.2.1 \
    torchvision==0.17.1

# Build Detectron2 against that exact PyTorch installation
RUN python -m pip install --no-cache-dir --no-build-isolation \
    'git+https://github.com/facebookresearch/detectron2.git'

COPY . .

RUN python -c "import torch; import torchvision; from detectron2 import _C; print(torch.__version__); print(torchvision.__version__); print('Detectron2 _C OK')"

CMD ["python", "your_application.py"]
```

Replace `your_application.py` with the actual entry point.

### 7. One more important point about your M1 Mac

Your host is Apple Silicon, but Docker is running a Linux environment. The `_C` library shown in the error is explicitly:

```text
aarch64-linux-gnu
```

That means the failing binary is a Linux ARM64 native extension, not the macOS ARM64 binary that worked on your host.

So you should **not assume that a Detectron2 package that was built and tested on the Mac can simply be reused inside Linux Docker**. The Linux container needs a compatible Linux ARM64 build of the extension.

This is especially important when publishing a prebuilt Detectron2 package to an internal/packagecloud repository. The package must be built for the target Python ABI and Linux architecture, and against the exact PyTorch binary used at runtime.

### 8. If it still fails

Compare the following from inside the container:

```bash
python -c "import torch; print(torch.__version__); print(torch.version.cuda); print(torch.__file__)"
python -c "import torchvision; print(torchvision.__version__); print(torchvision.__file__)"
python -m pip show detectron2
python -m detectron2.utils.collect_env
uname -m
```

Also check that your dependency installation is not doing this:

```text
install torch 2.2.1
install/build detectron2
install another requirements file
upgrade/downgrade torch
run application
```

That sequence can leave `_C.so` compiled against one PyTorch version while the container loads another one.

The correct order is:

```text
1. Choose the target architecture
2. Install the exact PyTorch + torchvision pair
3. Install/build Detectron2
4. Do not change PyTorch afterwards
5. Test `from detectron2 import _C`
6. Start the application
```

This matches Detectron2's documented guidance for undefined PyTorch/C++ symbols: the extension and PyTorch must come from a compatible build, and Detectron2 should be rebuilt after changing PyTorch. ([[GitHub](https://github.com/facebookresearch/detectron2/blob/main/INSTALL.md?utm_source=chatgpt.com)][1])

If you can share the output of `python -m detectron2.utils.collect_env` from inside the container, that should make it possible to identify exactly which dependency/build mismatch is causing the `_C.so` failure.

```

This is technically stronger because it does **not claim certainty about the exact mismatch** without seeing the container's environment. The documented root cause is a Detectron2/torch(vision) binary mismatch, while the `aarch64` path makes the Linux ARM64 build an additional thing to verify. :contentReference[oaicite:6]{index=6}
```

[1]: https://github.com/facebookresearch/detectron2/blob/main/INSTALL.md?utm_source=chatgpt.com "detectron2/INSTALL.md at main · facebookresearch/detectron2 · GitHub"
[2]: https://docs.pytorch.org/get-started/previous-versions/?utm_source=chatgpt.com "Previous PyTorch Versions | PyTorch"
